test/api/model/submessage_test.dart

```diff
@@ -111,4 +116,176 @@ void main() {
+        sender: eg.otherUser,
+        submessages: [
+          eg.submessage(
+            content: content,
+            senderId: eg.otherUser.userId,
+          ),
+          for (final (senderId, event) in events)
+            eg.submessage(
+              content: event,
+              senderId: senderId,
+            ),
+        ]);
+    }
+
+    test('parse poll from message', () {
+      check(messageWithSubmessageContents(pollWidgetData))
+        .poll.isNotNull()
+          ..question.equals('favorite letter')
+          ..options.deepEquals(defaultOptions);
+    });
+
+    test('parse poll with new question event', () {
+      check(messageWithSubmessageContents(
+        pollWidgetData,
+        events: [
+          (eg.otherUser.userId, PollQuestionEventSubmessage(question: 'new question')),
+        ],
+      ))
+        .poll.isNotNull()
+          ..question.equals('new question')
+          ..options.deepEquals(defaultOptions);
+    });
+
+    test('parse poll with new option event', () {
+      check(messageWithSubmessageContents(
+        pollWidgetData,
+        events: [
+          (eg.otherUser.userId, PollNewOptionEventSubmessage(idx: 3, option: 'D')),
+          (eg.selfUser.userId, PollNewOptionEventSubmessage(idx: 0, option: 'E')),
+        ],
+      ))
+        .poll.isNotNull()
+          ..question.equals('favorite letter')
+          ..options.deepEquals([
+            ...defaultOptions,
+            PollOption(text: 'D'),
+            PollOption(text: 'E'),
+          ]);
+    });
+
+    test('parse poll with vote events on initial canned options', () {
+      check(messageWithSubmessageContents(
+        pollWidgetData,
+        events: [
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: 'canned,1', op: PollVoteOp.add)),
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: 'canned,2', op: PollVoteOp.add)),
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: 'canned,2', op: PollVoteOp.remove)),
+          (eg.selfUser.userId, PollVoteEventSubmessage(key: 'canned,1', op: PollVoteOp.add)),
+        ],
+      ))
+        .poll.isNotNull()
+          ..question.equals('favorite letter')
+          ..options.deepEquals([
+            PollOption(text: 'A'),
+            eg.pollOption(text: 'B', voters: [eg.otherUser.userId, eg.selfUser.userId]),
+            PollOption(text: 'C'),
+          ]);
+    });
+
+    test('parse poll with vote events on post-creation options', () {
+      check(messageWithSubmessageContents(
+        pollWidgetData,
+        events: [
+          (eg.otherUser.userId, PollNewOptionEventSubmessage(idx: 0, option: 'D')),
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: '${eg.otherUser.userId},0', op: PollVoteOp.add)),
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: '${eg.selfUser.userId},0', op: PollVoteOp.add)),
+          (eg.otherUser.userId, PollVoteEventSubmessage(key: '${eg.selfUser.userId},0', op: PollVoteOp.remove)),
```

Comment:
Does this latter key correspond to an option? It looks like there's the initial options, then one created by eg.otherUser, and none by eg.selfUser.